src/ray/pubsub/subscriber.cc

```diff
@@ -227,14 +263,85 @@ void Subscriber::HandleLongPollingResponse(const rpc::Address &publisher_address
   }
 }
 
+void Subscriber::SendCommandBatchIfPossible(const rpc::Address &publisher_address) {
+  const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
+  auto command_batch_sent_it = command_batch_sent_.find(publisher_id);
+
+  // If there's no in flight command batch request to the publisher,
+  // send it. Otherwise, this function will be called when the in flight
+  // request is replied.
+  if (command_batch_sent_it == command_batch_sent_.end()) {
+    // Obtain the command queue.
+    auto command_queue_it = commands_.find(publisher_id);
+    if (command_queue_it == commands_.end()) {
+      return;
+    }
+    auto &command_queue = command_queue_it->second;
+
+    // Update the command in the FIFO order.
+    rpc::PubsubCommandBatchRequest command_batch_request;
+    command_batch_request.set_subscriber_id(subscriber_id_.Binary());
+    int64_t updated_commands = 0;
+    while (!command_queue.empty() && updated_commands < max_command_batch_size_) {
+      auto new_command = command_batch_request.add_commands();
+      new_command->Swap(command_queue.front().get());
+      command_queue.pop();
+      updated_commands += 1;
+    }
+
+    if (command_queue.size() == 0) {
+      commands_.erase(command_queue_it);
+    }
+
+    if (updated_commands == 0) {
+      return;
+    }
+
+    command_batch_sent_.emplace(publisher_id);
+    auto publisher_client = publisher_client_pool_.GetOrConnect(publisher_address);
+    publisher_client->PubsubCommandBatch(
+        command_batch_request,
+        [this, publisher_address, publisher_id](
+            Status status, const rpc::PubsubCommandBatchReply &reply) {
+          absl::MutexLock lock(&mutex_);
+          auto command_batch_sent_it = command_batch_sent_.find(publisher_id);
+          RAY_CHECK(command_batch_sent_it != command_batch_sent_.end());
+          command_batch_sent_.erase(command_batch_sent_it);
+          if (!status.ok()) {
+            // This means the publisher has failed.
+            // The publisher dead detection & command clean up will be done
+            // from the long polling request.
+            RAY_LOG(DEBUG) << "The command batch request to " << publisher_id
+                           << " has failed";
+          } else {
+            SendCommandBatchIfPossible(publisher_address);
+          }
+        });
+  }
+}
+
+void Subscriber::QueueCommand(const PublisherID &publisher_id,
+                              std::unique_ptr<rpc::Command> command) {
+  auto commands_it = commands_.find(publisher_id);
+  if (commands_it == commands_.end()) {
+    commands_.emplace(publisher_id, CommandQueue());
+  }
+  commands_[publisher_id].push(std::move(command));
```

Comment:
commands_[publisher_id].emplace(std::move(command));

Reply:
You mean .push right? commands is <id, std::queue>